### PR TITLE
GPU Task Runner message loop can be reconfigured

### PR DESCRIPTION
**Pull request by iskakaushik**
- Introduced `MessageLoopReconfigurableTaskRunner`, which
  can alternate between two message loops.

- GPU Task runner is now `MessageLoopReconfigurableTaskRunner`, this should maintain the exact same semantics when we have both task runners be the same.

- Coming up: engine to call `SwitchMessageLoop` on changes to layer tree where platform view is wither introduces for the first time or for the frame after which the platform view layer has been removed from the layer tree.

### CODE REVIEW

File: fml/msg_loop_reconfigurable_task_runner.cc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    return false;
+  }
+  fml::SharedLock lock(*shared_mutex_);
+  return MessageLoop::GetCurrent().GetLoopImpl() == loops_[current_loop_];
+}
+
+void MsgLoopReconfigurableTaskRunner::SwitchMessageLoop() {
+  fml::UniqueLock lock(*shared_mutex_);
+  fml::AutoResetWaitableEvent latch;
+  loops_[current_loop_]->PostTask([&latch]() { latch.Signal(); },
```
Comment (chinmaygarde) [inline fml/msg_loop_reconfigurable_task_runner.cc]: This by itself is not sufficient to also make sure timed tasks are completed on the right thread. Timed tasks that wanted to access components where they live in the switched configuration will get completed on the stale thread configuration.

File: fml/msg_loop_reconfigurable_task_runner.cc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  }
+  auto loop_1 = task_runner_1->loop_;
+  auto loop_2 = task_runner_2->loop_;
+  return MakeRefCounted<MsgLoopReconfigurableTaskRunner>(loop_1, loop_2);
+}
+
+MsgLoopReconfigurableTaskRunner::MsgLoopReconfigurableTaskRunner(
+    const fml::RefPtr<MessageLoopImpl>& loop_1,
+    const fml::RefPtr<MessageLoopImpl>& loop_2)
+    : TaskRunner(nullptr /* loop implemenation*/),
```
Comment (chinmaygarde) [inline fml/msg_loop_reconfigurable_task_runner.cc]: To address the issue of timed tasks finishing on the correct thread (described below), I suggest moving this switching logic to the message loop impl instead. You can create a subclass that itself takes two MessageLoopImpls and switches the thread assignment on the two when the thread configuration is changed. That way, timed tasks posted on one message loop will get executed on expiry on the new thread. You will be able to do this without latching also.

File: fml/task_runner_unittests.cc [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  runner->PostTask([&task_runner, &latch]() {
+    ASSERT_TRUE(task_runner->RunsTasksOnCurrentThread());
+    latch.CountDown();
+  });
+
+  latch.Wait();
+}
+
+TEST(MsgLoopReconfigurableTaskRunner, SwitchLoopBlocks) {
```
Comment (chinmaygarde) [inline fml/task_runner_unittests.cc]: Please add another test here for timed tasks.

File: fml/msg_loop_reconfigurable_task_runner.cc (hunk 2 above)
Comment (chinmaygarde) [reply fml/msg_loop_reconfigurable_task_runner.cc]: Another advantage of this approach could be that you dont need to change the public API of the shell and its components to vend these new `MsgLoopReconfigurableTaskRunner `s. Instead, you can keep the same task runner interface stable.

File: common/task_runners.cc [hunk 4]
```diff
@@ -13,6 +13,18 @@ TaskRunners::TaskRunners(std::string label,
                          fml::RefPtr<fml::TaskRunner> gpu,
                          fml::RefPtr<fml::TaskRunner> ui,
                          fml::RefPtr<fml::TaskRunner> io)
+    : label_(std::move(label)),
+      platform_(std::move(platform)),
+      gpu_(fml::MsgLoopReconfigurableTaskRunner::CreateFromSingleTaskRunner(
```
Comment (chinmaygarde) [inline common/task_runners.cc]: Lets call this `MergeableTaskRunner`

File: fml/message_loop_impl.cc [hunk 5]
```diff
@@ -119,8 +123,18 @@ void MessageLoopImpl::SwapTaskQueues(const fml::RefPtr<MessageLoopImpl>& other)
 
   std::lock(t1, t2, o1, o2, d1, d2);
 
-  std::swap(task_observers_, other->task_observers_);
-  std::swap(delayed_tasks_, other->delayed_tasks_);
+  // inherit all the task_observers.
```
Comment (chinmaygarde) [inline fml/message_loop_impl.cc]: We wont ever be able to unmerge without also splitting observers correctly. Lets just fail in this case.

File: fml/msg_loop_reconfigurable_task_runner.cc [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  if (merged_) {
+    return;
+  }
+  loops_[1]->InheritAllTasks(loops_[0]);
+  merged_ = true;
+}
+
+void MsgLoopReconfigurableTaskRunner::UnMergeLoops() {
+  fml::UniqueLock lock(*shared_mutex_);
+  merged_ = false;
```
Comment (chinmaygarde) [inline fml/msg_loop_reconfigurable_task_runner.cc]: Tasks like the AwaitVsync call will still be serviced on the old message loop. All you have done here is tell the message loop that new tasks for the rasterizer need to be serviced on the new configuration. Old awaits will now wake up the rasterizer on the wrong threads.

**Comment by cbracken** [comment]
@iskakaushik is this still WIP?

**Pull request closed without merging**